packages/mock/src/timestamp/tsa.ts

```diff
@@ -0,0 +1,178 @@
+/*
+Copyright 2023 The Sigstore Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { Crypto } from '@peculiar/webcrypto';
+import * as asn1js from 'asn1js';
+import * as pkijs from 'pkijs';
+import {
+  DIGEST_SHA256,
+  KEY_ALGORITHM_ECDSA_P256,
+  SIGNING_ALGORITHM_ECDSA_SHA384,
+} from '../constants';
+import { createRootCertificate } from '../util/root-cert';
+
+const ISSUER = 'CN=tsa,O=sigstore.mock';
+const OID_TSTINFO_CONTENT_TYPE = '1.2.840.113549.1.9.16.1.4';
+const OID_SIGNED_DATA_CONTENT_TYPE = '1.2.840.113549.1.7.2';
+
+export interface TSA {
+  rootCertificate: Buffer;
+  timestamp: (req: TimestampRequest) => Promise<Buffer>;
+}
+
+export interface TimestampRequest {
+  artifactHash: Buffer;
+  hashAlgorithmOID: string;
+  nonce: number;
+  policyOID: string;
+  certReq?: boolean;
+}
+
+export async function initializeTSA(): Promise<TSA> {
+  const root = await createRootCertificate(
+    ISSUER,
+    KEY_ALGORITHM_ECDSA_P256,
+    SIGNING_ALGORITHM_ECDSA_SHA384
+  );
+  return new TSAImpl({
+    rootCertificate: pkijs.Certificate.fromBER(root.cert.rawData),
+    keyPair: root.keyPair,
+  });
+}
+
+interface TSAOptions {
+  rootCertificate: pkijs.Certificate;
+  keyPair: CryptoKeyPair;
+}
+
+class TSAImpl implements TSA {
+  private rootCert: pkijs.Certificate;
+  private keyPair: CryptoKeyPair;
+  private crypto: pkijs.ICryptoEngine;
+  constructor(options: TSAOptions) {
+    this.rootCert = options.rootCertificate;
+    this.keyPair = options.keyPair;
+    this.crypto = new pkijs.CryptoEngine({ crypto: new Crypto() });
+  }
+
+  public get rootCertificate(): Buffer {
+    return Buffer.from(this.rootCert.toSchema().toBER(false));
+  }
+
+  public async timestamp(req: TimestampRequest): Promise<Buffer> {
```

Comment:
Is this basically implementing the bare bones of the tsa rfc? Worth linking to any of it?